src/service/AssignmentService.ts

```diff
@@ -18,27 +20,66 @@ const question = new Question(
     `SELECT * FROM Employees`
 );
 
-function compareResults(a: Array<any>, b: Array<any>) {
-    if (a.length !== b.length) return false;
+function compareResults(a: QueryResult, b: QueryResult) {
+    if (a.rows.length !== b.rows.length) return false;
 
     return JSON.stringify(a) === JSON.stringify(b);
 }
 
+function formatQueryResult(result: QueryResult) {
+    return result.success ? result.rows : { "error": result.error };
+}
+function formatSingleQueryResult(result: QueryResult) {
+    return result.success
+        ? (result.rows.length > 0 ? result.rows[0] : { "error": "No data was found" })
+        : { "error": result.error };
+}
+
 export function getQuestion() {
     return {prompt: question.prompt, context: question.context};
 }
 
-export function checkAnswer(userQuery: string) {
-    let result;
+export async function getAssignmentList(db: DatabaseConnection) {
+    const result: QueryResult = await db.exec(`
+        SELECT
+                a.id, a.title,
+                COUNT(q.id) 'questions',
+                COALESCE(SUM(q.points), 0) 'points'
+            FROM assignments a
+                LEFT JOIN questions q ON q.assignment_id = a.id
+            GROUP BY a.id, a.title;`);
+
+    return formatQueryResult(result);
+}
+
+export async function getQuestionList(db: DatabaseConnection, assignmentId: number) {
+    const result: QueryResult = await db.exec(`
+        SELECT id, question, points
+            FROM questions
+            WHERE assignment_id = ${assignmentId};`);
```

Comment:
All of these SELECTs are vulnerable to SQL injection attacks. We don't want client inputs to be passed directly into a SQL query under any circumstances. I believe there's a query function on db that can sanitize the inputs. I think the syntax would be something like: db.query(SELECT * FROM Table WHERE id = $1;, [id]);

Reply:
The ternary expressions in Index.ts were intended to address this, but handling this in the functions that execute the queries would probably be better. I'll look into this.

Reply:
Yeah you're correct about that, I just don't think the controller is the best place to handle it. Easy to imagine a future team expecting it to be handled in the DB layer (that's where I'd expect it anyways) and introducing a new API route that calls the function but doesn't protect against SQL injection since they might expect the DB function to handle the encoding.